### PR TITLE
ts: allow time interval and series to dump to be specified 

### PR DESCRIPTION
**Pull request by tbg**
First two commits from https://github.com/cockroachdb/cockroach/pull/50702, ignore them here.

----

Previously, only *all* of the timeseries could be dumped.
This is a lot of data. With this commit, we can pull any
subset of timeseries restricted to a time interval of
interest.

The plan is to use this from `debug.zip` and in `roachtest`,
for visualization via `./scripts/localmetrics`.

cc @lunevalex 

Release note: None

### CODE REVIEW

**Review by knz — COMMENTED** [review]
Sorry for the late review.
This work is still useful, as we got increasing requests to automate the extraction of ts data.

> Reviewed 2 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50728#-:-MHa7cBs7IR3zLWllKca:byfkqtj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @petermattis, and @tbg)

___
*[pkg/cli/tsdump.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50728#-MHa7KlRAAhHX66gfh6N:-MHa7KlS7DU4mjp2oPRt:b8bpz9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/9669df8ad346fe970fcdbe0a42edb6c00f65b847/pkg/cli/tsdump.go#L36)):*
> ```Go
> 		defer cancel()
> 
> 		var w tsWriter
> ```

Maybe add a TODO here to instead revert the flow of control and make the tsClient.Dump() code implement `rowStrIter` (from `format_table.go`). This way you can pass it to `printQueryOutput` and automatically benefit from all the formatters.
___
*[pkg/cli/tsdump.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50728#-MH_nAtjARh88iclx2qm:-MH_nAtkEFvW3-_Qv2OM:bjajcja) ([raw file](https://github.com/cockroachdb/cockroach/blob/9669df8ad346fe970fcdbe0a42edb6c00f65b847/pkg/cli/tsdump.go#L44)):*
> ```Go
> 			cw.w.Comma = '\t'
> 			w = cw
> 		default:
> ```

can you maybe add a special format type for your custom writer, and then make `default` print out an error that the format is unsupported
thanks
___
*[pkg/cli/tsdump.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50728#-MHa73WIBBYpoVMdoG21:-MHa73WJ3ygZt20NDUsL:bonhks3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9669df8ad346fe970fcdbe0a42edb6c00f65b847/pkg/cli/tsdump.go#L57)):*
> ```Go
> 		stream, err := tsClient.Dump(context.Background(), &tspb.DumpRequest{})
> 		if err != nil {
> 			log.Fatalf(context.Background(), "%v", err)
> ```

for client commands we don't typically fatal but return instead
___
*[pkg/cli/tsdump.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50728#-MH_nL35AapN9xu-nHRh:-MH_nL35AapN9xu-nHRi:b-f4g6x9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9669df8ad346fe970fcdbe0a42edb6c00f65b847/pkg/cli/tsdump.go#L63)):*
> ```Go
> 			data, err := stream.Recv()
> 			if err == io.EOF {
> 				return w.Flush()
> ```

I'd defer the `w.Flush` at the beginning instead, with an `errors.CombineErrors()` call int he defer

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, 7 of 7 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50728)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @tbg)

File: pkg/cli/tsdump.go [hunk 1]
```diff
@@ -51,8 +53,15 @@ Dumps all of the raw timeseries values in a cluster.
...
 
+		resp, err := serverpb.NewAdminClient(conn).ChartCatalog(ctx, &serverpb.ChartCatalogRequest{})
+		if err != nil {
+			return err
+		}
+
 		tsClient := tspb.NewTimeSeriesClient(conn)
-		stream, err := tsClient.Dump(context.Background(), &tspb.DumpRequest{})
+		stream, err := tsClient.Dump(context.Background(), &tspb.DumpRequest{
+			Names: catalog.AllMetricsNames(resp.Catalog...),
```
Comment (tbg) [inline pkg/cli/tsdump.go:63]: This should be filled in on the server, so that we use the server's catalog (matters when the cli is old)

Comment (tbg) [reply pkg/cli/tsdump.go:63]: Nevermind, I was reading this wrong. This is already pulling them from the server so all good

File: pkg/ts/server.go [hunk 2]
```diff
@@ -308,37 +308,69 @@ func (s *Server) Query(
...
+			if err := dumpTimeseriesAllSources(db, seriesName, res, req.StartNanos, req.EndNanos, stream); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func dumpTimeseriesAllSources(
+	db *kv.DB,
```
Comment (tbg) [inline pkg/ts/server.go:338]: replace w/ interface

**Comment by petermattis** [comment]
@tbg Do you have any intention of resuscitating this PR? Perhaps convert it to a draft in the meantime.

**Comment by tbg** [comment]
I met with @vrongmeal (contributor) today, he is planning to work on it, though not in this PR.

**Pull request closed without merging**